columnflow/tasks/framework/base.py

```diff
@@ -1213,11 +1194,123 @@ def resolve_param_values(cls, params: dict) -> dict:
                 if "config_insts" not in params and "configs" in params:
                     params["config_insts"] = list(map(analysis_inst.get_config, params["configs"]))
 
-            # for backwards compatibility, we automatically switch the defaults from config to analysis inst
-            params = cls.switch_to_analysis_inst_aux(params)
+        # resolving of parameters that is required before ArrayFunctions etc. can be initialized
+        params = cls.resolve_pre_taf_init_params(params)
+
+        # initialize ArrayFunctions etc. and collect known shifts
+        shifts = TaskShifts()
+        params = cls.build_taf_insts(params, shifts)
+
+        # resolving of parameters that can only be performed after ArrayFunction initialization
+        params = cls.resolve_post_taf_init_params(params)
+
+        # resolving of shifts
+        params = cls.resolve_shifts(params, shifts)
+
+        return params
+
+    @classmethod
+    def build_taf_insts(cls, params: dict[str, Any], shifts: TaskShifts) -> dict[str, Any]:
+        """
+        Build the array function instances.
+        For single-config/dataset tasks, build_taf_insts is implemented by mixin classes such as the ProducersMixin.
+        For multi-config tasks, build_taf_insts from the upstream task is called for each config instance.
+        If the build_taf_insts function needs to be called for other combinations of parameters (e.g. per dataset),
+        it can be overwritten by the task class.
+
+        :param params: Dictionary of task parameters.
+        :param shifts: Collection of local and global shifts.
+        :return: Updated dictionary of task parameters.
+        """
+
+        # check if shifts are already known
+        if params.get("known_shifts", None) and params.get("branch", -1) != -1:
```

Comment:
Just a reminder for us to confine the caching of known shifts to the interaction between workflows and branches using law's `req_branch`.